File: src/etch/bindings/csharp/transport/test/TestMailboxManager.cs
using System;
using System.Collections.Generic;
using System.Text;
using Etch.Msg;
using Etch.Support;
using NUnit.Framework;
using System.Net;

namespace Etch.Transport.Test
{
    [TestFixture]
    public class TestMailboxManager
    {
        private MailboxManager mmgr;
        private Message msg;
        // private DefaultValueFactory vf = new DefaultValueFactory();
        private MyMessageHandler mh;
        private MyMessagizer m;

        private bool duplexFlag = true;
        private bool registerFlag = false;

        private readonly ValueFactory vf = new MyValueFactory();

        public TestMailboxManager() 
        { 
            //mh = new MyMessageHandler(this);
            //m = new MyMessagizer( this );
        }

        public Message ConstructSimpleMessage()
        {
            Message msg = new Message( MyValueFactory.mt_function, vf );
            return msg;
        }

        public Message ConstructAddMessage() 
	    {
		    msg = new Message( MyValueFactory.mt_add, vf );
		    msg.Add( MyValueFactory.mf_x, 4 );
		    msg.Add( MyValueFactory.mf_y, 5 );
    		
		    return msg;
	    }

        public void SetUp()
        {
            mh = new MyMessageHandler( this );
            
            msg = ConstructAddMessage();
            mmgr = new MailboxManager( mh, 1000 );
            m = new MyMessagizer( this, mmgr );
            mmgr.SetSource( m );
        }

        //[Test]
        //public void Up() 
        //{
        //    SetUp();
        //    Assert.IsNull(mh.what);
        //    mmgr.Up(m);
        //    Assert.AreEqual(mh.what, What.MMGRMESSAGEHANDLERUP);
        //}

        //[Test]
        //public void Down() 
        //{
        //    SetUp();
        //    Assert.IsNull(mh.what);
        //    mmgr.Down(m);
        //    Assert.AreEqual(mh.what, What.MMGRMESSAGEHANDLERDown);
        //}

        //[Test]
        //public void Close() 
        //{
        //    SetUp();
        //    Assert.IsNull(m.what);
        //    mmgr.Close(true);
        //    Assert.AreEqual(m.what, What.MYMESSAGIZERClose);
        //}
    	
        //[Test]
        //public void LocalAddress() 
        //{
        //    SetUp();
        //    Assert.IsNull(m.what);
        //    mmgr.LocalAddress();
        //    Assert.AreEqual(m.what, What.MYMESSAGIZERLocalAddress);
        //}
    	
        //[Test]
        //public void RemoteAddress() 
        //{
        //    SetUp();
        //    Assert.IsNull(m.what);
        //    mmgr.RemoteAddress();
        //    Assert.AreEqual(m.what, What.MYMESSAGIZERRemoteAddress);
        //}
    	
        //[Test]
        //public void ShutdownInput() 
        //{
        //    SetUp();
        //    Assert.IsNull(m.what);
        //    mmgr.ShutdownInput();
        //    Assert.AreEqual(m.what, What.MYMESSAGIZERShutdownInput);
        //}
    	
        //[Test]
        //public void ShutdownOutput() 
        //{
        //    SetUp();
        //    Assert.IsNull(m.what);
        //    mmgr.ShutdownOutput();
        //    Assert.AreEqual(m.what, What.MYMESSAGIZERShutdownOutput);
        //}
    	
        //[Test]
        //public void Stop() 
        //{
        //    SetUp();
        //    Assert.IsNull(m.what);
        //    mmgr.Stop();
        //    Assert.AreEqual(m.what, What.MYMESSAGIZERSTOP);
        //}
    	
	    [Test]
	    public void SimplexMessage() 
	    {
            SetUp();
		    Assert.IsNull(m.what);
    		
		    Message message = ConstructSimpleMessage();
		    mmgr.Messagex( null, message );
    		
		    Assert.AreEqual(m.what, What.MYMESSAGIZERMESSAGE);
	    }
    	
	    [Test]
	    public void DuplexMessage() 
	    {
            SetUp();
		    Assert.IsNull(m.what);
		    Assert.IsNull(mh.what);
    		
		    Message message = ConstructAddMessage();
            mmgr.Message( null, message, 1, 0 );
    		
		    Assert.AreEqual(m.what, What.MYMESSAGIZERMESSAGE);
    		
		    /* 
		     * Because PlainMailbox will intercept messages ...
		     * The Default messagehandler won't be called.
		     */
		    Assert.IsNull(mh.what);
    		
    		
	    }
    	
	    [Test]
	    public void DuplicateRegistration() 
	    {
            SetUp();
		    Message message = ConstructAddMessage();
		    Mailbox mb = 
                ( Mailbox ) mmgr.Message( null, message, 1, 0 );

		    /*
		     * Check for duplicate registration
		     */
		    try 
		    {
			    mmgr.Register(mb);
		    }
		    catch (Exception)
		    {
			    registerFlag = true;
		    }
		    Assert.IsTrue(registerFlag);
    		
	    }
    	
	    [Test]
	    public void TestUnregistration() 
	    {
            SetUp();
		    Message message = ConstructAddMessage();
		    Mailbox mb =  mmgr.Message( null, message, 1, 0 );

		    /*
		     * Unregister and then register again ... 
		     */
		    mmgr.Unregister(mb);
            registerFlag = false;
		    try 
		    {
                mmgr.Register( ( Mailbox ) mb );
		    }
		    catch (Exception)
		    {
                registerFlag = true;
		    }
		    Assert.IsFalse(registerFlag);
    	
	    }
    	
	    [Test]
	    public void NoMailboxOrReplyToId() 
	    {
            SetUp();
		    Assert.IsNull(mh.what);
		    Assert.IsNull(m.what);
    		
		    duplexFlag = false;
		    Message message = ConstructAddMessage();
    		
		    m.Messagex(null, message);
    		
		    Assert.AreEqual(m.what, What.MYMESSAGIZERMESSAGE);
		    /*
		     * Since no mailbox, default message handler receives
		     * messages.
		     */
		    Assert.AreEqual(mh.what, What.MMGRMESSAGEHANDLERMESSAGE);
	    }

        public enum What
        {
            MMGRMESSAGEHANDLERSTARTED,
            MMGRMESSAGEHANDLERUP,
            MMGRMESSAGEHANDLERDown,
            MMGRMESSAGEHANDLERSTOPPED,
            MYMESSAGIZERClose,
            MYMESSAGIZERLocalAddress,
            MYMESSAGIZERRemoteAddress,
            MYMESSAGIZERShutdownInput,
            MYMESSAGIZERShutdownOutput,
            MYMESSAGIZERSTOP,
            MYMESSAGIZERMESSAGE,
            MMGRMESSAGEHANDLERMESSAGE, 
            OOB_CONTROL_HANDLER, 
            OOB_NOTIFY_HANDLER, 
            OOB_QUERY_HANDLER
        };

        public class MyMessageHandler : MessageHandler<MailboxMessageSource>
        {
            public Enum what;

            public Who xsender;

            public Message xmsg;

            public Object xcontrol;

            public Object xvalue;

            public Object xquery;

            public Object xevent;

            public Object xquery_result;

            public TestMailboxManager _tmm;

            public MyMessageHandler()
            {
                // nothing to do.
            }

            public MyMessageHandler( TestMailboxManager tmm )
            {
                _tmm = tmm;
            }
    		
		    public bool Message( Who sender, Message msg ) 
		    {
			    what = What.MMGRMESSAGEHANDLERMESSAGE;
                xsender = sender;
                xmsg = msg;
                return true;
		    }

            #region SourceHandler<MailboxMessageSource> Members

            public MailboxMessageSource GetSource()
            {
                // ignore
                return null;
            }

            public void SetSource( MailboxMessageSource src )
            {
                // ignore
            }

            #endregion

            #region Session Members

            public object SessionQuery( object query )
            {
                what = What.OOB_QUERY_HANDLER;
                xquery = query;
                return xquery_result;
            }

            public void SessionControl( object control, object value )
            {
                what = What.OOB_CONTROL_HANDLER;
                xcontrol = control;
                xvalue = value;
            }

            public void SessionNotify( object eventObj )
            {
                what = What.OOB_NOTIFY_HANDLER;
                xevent = eventObj;
            }

            #endregion
        }

        public class MyMessagizer : PacketHandler, MessagizerMessageSource
        {
            private MessageHandler handler;
    		
		    public Enum what;

            public TestMailboxManager _tmm;

            public MyMessagizer( TestMailboxManager tmm )
            {
                _tmm = tmm;
            }

            public MyMessagizer( TestMailboxManager tmm, MessageHandler handler)
            {
                _tmm = tmm;
                this.handler = handler;
            }

            public override string ToString()
            {
                return "My Messagizer";
            }

            public void Packet( PacketSource src, Who sender, FlexBuffer buf ) 
		    {
			    
			    // ignore
		    }

            public void Messagex( Who recipient, Message msg ) 
		    {
			    // message from mailbox manager
			    what = What.MYMESSAGIZERMESSAGE;
    			
			    // send message back ... 
                if ( _tmm.duplexFlag == true )
                    msg.InReplyTo = msg.MessageId;
				    
			    handler.Message(null, msg);
		    }

            #region PacketHandler Members

            public void Packet( Who sender, FlexBuffer buf )
            {
                throw new Exception( "The method or operation is not implemented." );
            }

            #endregion

            #region SourceHandler<PacketSource> Members

            public PacketSource GetSource()
            {
                return null;
            }

            public void SetSource( PacketSource src )
            {
                // ignore
            }

            #endregion

            #region Session Members

            public object SessionQuery( object query )
            {
                return null;
            }

            public void SessionControl( object control, object value )
            {
                // ignore
            }

            public void SessionNotify( object eventObj )
            {
                // ignore
            }

            #endregion

            #region Source Members

            public object GetHandler()
            {
                return null;
            }

            public void SetHandler( object handler )
            {
                // ignore
            }

            #endregion

            #region Transport Members

            public object TransportQuery( object query )
            {
                return null;
            }

            public void TransportControl( object control, object value )
            {
                // ignore
            }

            public void TransportNotify( object eventObj )
            {
                // ignore
            }

            #endregion
        }

        public class MyValueFactory : DefaultValueFactory
        {
            private readonly static TypeMap types = new TypeMap();

            private readonly static FieldMap fields = new FieldMap();

            public readonly static XType mt_function = types.Get( "function" );

            public readonly static XType mt_add = types.Get( "add" );

            public readonly static XType mt_add_result = types.Get( "add_result" );

            public readonly static Field mf_x = fields.Get( "x" );

            public readonly static Field mf_y = fields.Get( "y" );

            public readonly static Field mf_result = fields.Get( "xresult" );

            static MyValueFactory()
            {
                DefaultValueFactory.Init( types, fields );

                mt_function.PutValidator( DefaultValueFactory._mf__messageId, Validator_long.Get( 0 ) );
                mt_function.PutValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.Get( 0 ) );

                mt_add.PutValidator( mf_x, Validator_int.Get( 0 ) );
                mt_add.PutValidator( mf_y, Validator_int.Get( 0 ) );
                mt_add.PutValidator( DefaultValueFactory._mf__messageId, Validator_long.Get( 0 ) );
                mt_add.PutValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.Get( 0 ) );

                mt_add_result.PutValidator( mf_result, Validator_int.Get( 0 ) );
                mt_add_result.PutValidator( DefaultValueFactory._mf__messageId, Validator_long.Get( 0 ) );
                mt_add_result.PutValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.Get( 0 ) );
            }

            public override XType GetType( int? id )
            {
                return types.Get( id );
            }

            public override XType GetType( string name )
            {
                return types.Get( name );
            }

            public override ICollection<XType> GetTypes()
            {
                return types.Values();
            }

            public override Field GetField( int? id )
            {
                return fields.Get( id );
            }

            public override Field GetField( string name )
            {
                return fields.Get( name );
            }

            public override ICollection<Field> GetFields()
            {
                return fields.Values();
            }
        }
    }
}

File: src/etch/bindings/java/transport/test/TestMailboxManager.java
package etch.bindings.java.transport.test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import metreos.core.Who;
import metreos.core.io.PacketHandler;
import metreos.core.io.PacketSource;
import metreos.util.FlexBuffer;

import org.junit.Before;
import org.junit.Test;

import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.Message;
import etch.bindings.java.msg.Type;
import etch.bindings.java.msg.Validator_int;
import etch.bindings.java.msg.Validator_long;
import etch.bindings.java.support.DefaultValueFactory;
import etch.bindings.java.support.Mailbox;
import etch.bindings.java.transport.MailboxManager;
import etch.bindings.java.transport.MailboxMessageSource;
import etch.bindings.java.transport.MessageHandler;
import etch.bindings.java.transport.MessagizerMessageSource;

public class TestMailboxManager 
{

	private MailboxManager mmgr;
	private Message msg;
//	private DefaultValueFactory vf = new DefaultValueFactory();
	private MyMessageHandler mh = new MyMessageHandler();
	private MyMessagizer m;
	
	/** eh */
	boolean duplexFlag = true;
	private boolean registerFlag = false;
	
	public Message constructSimpleMessage() throws Exception
	{
		Type mt_function = new Type( "function" );

		mt_function.putValidator( DefaultValueFactory._mf__messageId, Validator_long.get( 0 ) );
		mt_function.putValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.get( 0 ) );
		
		DefaultValueFactory vf = new DefaultValueFactory();
		vf.addType(mt_function);
		
		msg = new Message(mt_function, vf);
		return msg;
	}
	
	public Message constructAddMessage() throws Exception 
	{
		Type mt_add = new Type( "add" );
		Type mt_add_result = new Type( "add_result" );
		
		Field mf_x = new Field( "x" );
		Field mf_y = new Field( "y" );
		Field mf_result = new Field( "result" );

		mt_add.putValidator( mf_x, Validator_int.get( 0 ) );
		mt_add.putValidator( mf_y, Validator_int.get( 0 ) );
		mt_add.putValidator( DefaultValueFactory._mf__messageId, Validator_long.get( 0 ) );
		mt_add.putValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.get( 0 ) );
		
		mt_add_result.putValidator( mf_result, Validator_int.get( 0 ) );
		mt_add_result.putValidator( DefaultValueFactory._mf__messageId, Validator_long.get( 0 ) );
		mt_add_result.putValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.get( 0 ) );
		
		DefaultValueFactory vf = new DefaultValueFactory();
		vf.addType( mt_add );
		vf.addType( mt_add_result );
		vf.addField( mf_x );
		vf.addField( mf_y );
		vf.addField( mf_result );
		
		msg = new Message( mt_add, vf );
		msg.put( mf_x, 4 );
		msg.put( mf_y, 5 );
		
		return msg;
	}
	
	@Before
	public void setUp() throws Exception
	{
		msg = constructAddMessage();
		mmgr = new MailboxManager(mh, 1000);
		m = new MyMessagizer(mmgr);
		mmgr.setSource(m);
	}
	
//	@Test
//	public void up() throws Exception
//	{
//		assertNull(mh.what);
//		mmgr.up(m);
//		assertSame(mh.what, What.MMGRMESSAGEHANDLERUP);
//	}
	
//	@Test
//	public void down() throws Exception
//	{
//		assertNull(mh.what);
//		mmgr.down(m);
//		assertSame(mh.what, What.MMGRMESSAGEHANDLERDOWN);
//	}
	
//	@Test
//	public void close() throws Exception
//	{
//		assertNull(m.what);
//		mmgr.close(true);
//		assertSame(m.what, What.MYMESSAGIZERCLOSE);
//	}
	
//	@Test
//	public void localAddress() throws Exception
//	{
//		assertNull(m.what);
//		mmgr.localAddress();
//		assertSame(m.what, What.MYMESSAGIZERLOCALADDRESS);
//	}
	
//	@Test
//	public void remoteAddress() throws Exception
//	{
//		assertNull(m.what);
//		mmgr.remoteAddress();
//		assertSame(m.what, What.MYMESSAGIZERREMOTEADDRESS);
//	}
	
//	@Test
//	public void shutdownInput() throws Exception
//	{
//		assertNull(m.what);
//		mmgr.shutdownInput();
//		assertSame(m.what, What.MYMESSAGIZERSHUTDOWNINPUT);
//	}
	
//	@Test
//	public void shutdownOutput() throws Exception
//	{
//		assertNull(m.what);
//		mmgr.shutdownOutput();
//		assertSame(m.what, What.MYMESSAGIZERSHUTDOWNOUTPUT);
//	}
	
//	@Test
//	public void stop() throws Exception
//	{
//		assertNull(m.what);
//		mmgr.stop();
//		assertSame(m.what, What.MYMESSAGIZERSTOP);
//	}
	
	@Test
	public void simplexMessage() throws Exception
	{
		assertNull(m.what);
		
		Message message = constructSimpleMessage();
		mmgr.messagex( null, message );
		
		assertSame(m.what, What.MYMESSAGIZERMESSAGE);
	}
	
	@Test
	public void duplexMessage() throws Exception
	{
		
		assertNull(m.what);
		assertNull(mh.what);
		
		Message message = constructAddMessage();
		mmgr.message(null, message, 1, 0);
		
		assertSame(m.what, What.MYMESSAGIZERMESSAGE);
		
		/* 
		 * Because PlainMailbox will intercept messages ...
		 * The Default messagehandler won't be called.
		 */
		assertNull(mh.what);
		
		
	}
	
	@Test
	public void duplicateRegistration() throws Exception
	{
		Message message = constructAddMessage();
		Mailbox mb = mmgr.message(null, message, 1, 0);

		/*
		 * Check for duplicate registration
		 */
		try 
		{
			mmgr.register(mb);
		}
		catch (IllegalArgumentException e)
		{
			registerFlag = true;
		}
		assertTrue(registerFlag);
		
	}
	
	@Test
	public void testUnregistration() throws Exception
	{
		Message message = constructAddMessage();
		Mailbox mb = mmgr.message(null, message, 1, 0);

		/*
		 * Unregister and then register again ... 
		 */
		mmgr.unregister(mb);
		
		try 
		{
			mmgr.register(mb);
		}
		catch (IllegalArgumentException e)
		{
			registerFlag = true;
		}
		assertFalse(registerFlag);
	
	}
	
	@Test
	public void noMailboxOrReplyToId() throws Exception
	{
		
		assertNull(mh.what);
		assertNull(m.what);
		
		duplexFlag = false;
		Message message = constructAddMessage();
		
		m.messagex(null, message);
		
		assertSame(m.what, What.MYMESSAGIZERMESSAGE);
		/*
		 * Since no mailbox, default message handler receives
		 * messages.
		 */
		assertSame(mh.what, What.MMGRMESSAGEHANDLERMESSAGE);
	}
	
	public enum What {MMGRMESSAGEHANDLERSTARTED, MMGRMESSAGEHANDLERUP, MMGRMESSAGEHANDLERDOWN, 
		MMGRMESSAGEHANDLERSTOPPED, MYMESSAGIZERCLOSE, MYMESSAGIZERLOCALADDRESS, 
		MYMESSAGIZERREMOTEADDRESS, MYMESSAGIZERSHUTDOWNINPUT, MYMESSAGIZERSHUTDOWNOUTPUT, 
		MYMESSAGIZERSTOP, MYMESSAGIZERMESSAGE, MMGRMESSAGEHANDLERMESSAGE, OOB_CONTROL_HANDLER, OOB_NOTIFY_HANDLER, OOB_QUERY_HANDLER};
	
	public class MyMessageHandler implements MessageHandler<MailboxMessageSource>
	{
		public What what;
		
		public Who xsender;
		
		public Message xmsg;
		
		public Object xcontrol;
		
		public Object xvalue;
		
		public Object xquery;
		
		public Object xevent;
		
		public Object xquery_result;
		
		public MyMessageHandler() 
		{
			// nothing to do
		}
		
		public boolean message( Who sender, Message msg ) throws Exception
		{
			what = What.MMGRMESSAGEHANDLERMESSAGE;
			xsender = sender;
			xmsg = msg;
			return true;
		}

		public MailboxMessageSource getSource()
		{
			// ignore.
			return null;
		}

		public void setSource( MailboxMessageSource src )
		{
			// ignore.
		}

		public Object sessionQuery( Object query )
		{
			what = What.OOB_QUERY_HANDLER;
			xquery = query;
			return xquery_result;
		}

		public void sessionControl( Object control, Object value )
		{
			what = What.OOB_CONTROL_HANDLER;
			xcontrol = control;
			xvalue = value;
		}

		public void sessionNotify( Object event )
		{
			what = What.OOB_NOTIFY_HANDLER;
			xevent = event;
		}
	}
	
	public class MyMessagizer implements PacketHandler, MessagizerMessageSource 
	{
		private final MessageHandler handler;
		
		private PacketSource pktSrc;
		
		public What what;
		
		public MyMessagizer( MessageHandler handler )
		{
			this.handler = handler;
		}
		
		@Override
		public String toString()
		{
			return String.format( "myMessagizer/%s", pktSrc );
		}
		
		public void packet( Who sender, FlexBuffer buf ) throws Exception
		{
			// ignore.
		}
		
		
		public void messagex( Who recipient, Message msg ) throws Exception
		{
			// message from mailbox manager.
			
			what = What.MYMESSAGIZERMESSAGE;
			
			// send message back ... 
			if (duplexFlag == true)
				msg.setInReplyTo(msg.getMessageId());
			
			handler.message(null, msg);
		}

		public PacketSource getSource()
		{
			// TODO Auto-generated method stub
			return null;
		}

		public void setSource( PacketSource src )
		{
			// TODO Auto-generated method stub
			
		}

		public Object sessionQuery( Object query )
		{
			// TODO Auto-generated method stub
			return null;
		}

		public void sessionControl( Object control, Object value )
		{
			// TODO Auto-generated method stub
			
		}

		public void sessionNotify( Object event )
		{
			// TODO Auto-generated method stub
			
		}

		public MessageHandler<MessagizerMessageSource> getHandler()
		{
			// TODO Auto-generated method stub
			return null;
		}

		public void setHandler( MessageHandler<MessagizerMessageSource> handler )
		{
			// TODO Auto-generated method stub
			
		}

		public Object transportQuery( Object query ) throws Exception
		{
			// TODO Auto-generated method stub
			return null;
		}

		public void transportControl( Object control, Object value ) throws Exception
		{
			// TODO Auto-generated method stub
			
		}

		public void transportNotify( Object event ) throws Exception
		{
			// TODO Auto-generated method stub
			
		}
	}
}
